### PR TITLE
[Droidcon Italy] Counter and Dogs sample apps

### PR DESCRIPTION
**Pull request by gpeal**
This code was created for Droidcon Italy but could reasonably be committed to the repo as additional samples.

### CODE REVIEW

File: dogs/src/main/java/com/airbnb/mvrx/dogs/data/DogRepository.kt [hunk 1]
```diff
@@ -0,0 +1,76 @@
+@file:Suppress("Detekt.MaxLineLength")
+
+package com.airbnb.mvrx.dogs.data
+
+import io.reactivex.Observable
+import io.reactivex.Single
+
+class DogRepository {
+    fun getDogs() = Observable.fromCallable<List<Dog>> {
+        Thread.sleep(2000)
```
Comment (BenSchwab) [inline dogs/src/main/java/com/airbnb/mvrx/dogs/data/DogRepository.kt]: `fromCallable.delay(2000)`

File: dogs/src/main/java/com/airbnb/mvrx/dogs/data/DogRepository.kt [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+                "On Tue Apr 02 08:10:18 PDT 2019 Happy was available. For current animal profile, images and videos please visit http://petportal.oaklandanimalservices.org/myapp/publicAnimalProfile/show/35137 All animals are listed as spayed/neutered, even if they currently are not, since they will be fixed before leaving the shelter. To adopt: Other canines in the home must meet the new dog prior to adoption. Adoptable dogs can be “cat-tested” upon request. You must provide proof of home ownership or permission to have a dog from your landlord. We cannot hold the dog that you want to adopt while you try to meet this requirement. Adopters are encouraged to bring all the human members of their household, and must bring children age 12 and under residing in the house, to meet their new companion. For information on how to redeem a lost and found animal visit http://www.oaklandanimalservices.org/services/lost-and-found/redeeming-your-pet/ Please bring the animal's number with you to assist us in finding them in the shelter."
+            )
+        )
+    }
+
+    fun adoptDog(dog: Dog) = Single.fromCallable {
+        Thread.sleep(5000)
```
Comment (BenSchwab) [inline dogs/src/main/java/com/airbnb/mvrx/dogs/data/DogRepository.kt]: can also use delay

File: dogs/src/main/java/com/airbnb/mvrx/dogs/DogViewModel.kt [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+import io.reactivex.schedulers.Schedulers
+
+class DogViewModel(
+    state: DogState,
+    private val dogRepository: DogRepository
+) : MvRxViewModel<DogState>(state) {
+
+    init {
+        dogRepository.getDogs()
+            .subscribeOn(Schedulers.io())
```
Comment (hellohuanlin) [inline dogs/src/main/java/com/airbnb/mvrx/dogs/DogViewModel.kt]: Nit: move `subscribeOn(Schedulers.io())` to the end of `getDogs` function inside repository

File: dogs/src/main/java/com/airbnb/mvrx/dogs/DogViewModel.kt [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+            .execute { copy(dogs = it) }
+    }
+
+    fun loveDog(dogId: Long) = setState { copy(lovedDogId = dogId) }
+
+    fun adoptLovedDog() = withState { state ->
+        val lovedDog = state.lovedDog ?: throw IllegalStateException("You must love a dog first!")
+        dogRepository.adoptDog(lovedDog)
+            .subscribeOn(Schedulers.io())
```
Comment (hellohuanlin) [inline dogs/src/main/java/com/airbnb/mvrx/dogs/DogViewModel.kt]: same here. the data producer has full context of which thread the data need to be produced.

**Pull request merged**